File: modernized/bill-payment/src/models/transaction.ts
import { Money } from '../money';

/**
 * Transaction entity — modern equivalent of copybook `CVTRA05Y` (VSAM file
 * TRANSACT, record length 350). Field comments show the original COBOL PIC clause.
 */
export interface Transaction {
  /** TRAN-ID            PIC X(16) — 16-char key; numeric content, zero-padded */
  transactionId: string;
  /** TRAN-TYPE-CD       PIC X(02) */
  typeCode: string;
  /** TRAN-CAT-CD        PIC 9(04) */
  categoryCode: number;
  /** TRAN-SOURCE        PIC X(10) */
  source: string;
  /** TRAN-DESC          PIC X(100) */
  description: string;
  /** TRAN-AMT           PIC S9(09)V99 */
  amount: Money;
  /** TRAN-MERCHANT-ID   PIC 9(09) */
  merchantId: string;
  /** TRAN-MERCHANT-NAME PIC X(50) */
  merchantName: string;
  /** TRAN-MERCHANT-CITY PIC X(50) */
  merchantCity: string;
  /** TRAN-MERCHANT-ZIP  PIC X(10) */
  merchantZip: string;
  /** TRAN-CARD-NUM      PIC X(16) */
  cardNumber: string;
  /** TRAN-ORIG-TS       PIC X(26) — "YYYY-MM-DD HH:MM:SS.mmmmmm" */
  originTimestamp: string;
  /** TRAN-PROC-TS       PIC X(26) */
  processTimestamp: string;
}

/**
 * Constants used by COBIL00C when building the bill-payment transaction record.
 * These mirror the literal MOVEs in the PROCESS-ENTER-KEY paragraph.
 */
export const BILL_PAYMENT_TXN = {
  TYPE_CODE: '02',
  CATEGORY_CODE: 2,
  SOURCE: 'POS TERM',
  DESCRIPTION: 'BILL PAYMENT - ONLINE',
  MERCHANT_ID: '999999999',
  MERCHANT_NAME: 'BILL PAYMENT',
  MERCHANT_CITY: 'N/A',
  MERCHANT_ZIP: 'N/A',
} as const;

export const TRAN_ID_LENGTH = 16;

/** Format a numeric transaction id into the 16-char zero-padded `PIC X(16)` form. */
export function formatTransactionId(value: number): string {
  return value.toString().padStart(TRAN_ID_LENGTH, '0');
}

/**
 * Build a 26-char CICS-style timestamp ("YYYY-MM-DD HH:MM:SS.mmmmmm"),
 * equivalent to GET-CURRENT-TIMESTAMP (ASKTIME + FORMATTIME) in COBIL00C.
 */
export function buildTimestamp(now: Date = new Date()): string {
  const pad = (n: number, w = 2) => n.toString().padStart(w, '0');
  const date = `${now.getUTCFullYear()}-${pad(now.getUTCMonth() + 1)}-${pad(now.getUTCDate())}`;
  const time = `${pad(now.getUTCHours())}:${pad(now.getUTCMinutes())}:${pad(now.getUTCSeconds())}`;
  // COBIL00C sets the microseconds portion (WS-TIMESTAMP-TM-MS6) to zeros.
  return `${date} ${time}.000000`;
}

File: java/src/main/java/com/carddemo/model/entity/Transaction.java
package com.carddemo.model.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import java.math.BigDecimal;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

/**
 * COBOL copybook: CVTRA05Y (TRAN-RECORD), VSAM file TRANSACT, RECLN 350.
 */
@Entity
@Table(name = "transaction")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Transaction {

    /** TRAN-ID PIC X(16). */
    @Id
    @Column(name = "tran_id", length = 16, nullable = false)
    private String transactionId;

    /** TRAN-TYPE-CD PIC X(02). */
    @Column(name = "tran_type_cd", length = 2)
    private String typeCode;

    /** TRAN-CAT-CD PIC 9(04). */
    @Column(name = "tran_cat_cd")
    private Integer categoryCode;

    /** TRAN-SOURCE PIC X(10). */
    @Column(name = "tran_source", length = 10)
    private String source;

    /** TRAN-DESC PIC X(100). */
    @Column(name = "tran_desc", length = 100)
    private String description;

    /** TRAN-AMT PIC S9(09)V99. */
    @Column(name = "tran_amt", precision = 11, scale = 2)
    private BigDecimal amount;

    /** TRAN-MERCHANT-ID PIC 9(09). */
    @Column(name = "tran_merchant_id")
    private Long merchantId;

    /** TRAN-MERCHANT-NAME PIC X(50). */
    @Column(name = "tran_merchant_name", length = 50)
    private String merchantName;

    /** TRAN-MERCHANT-CITY PIC X(50). */
    @Column(name = "tran_merchant_city", length = 50)
    private String merchantCity;

    /** TRAN-MERCHANT-ZIP PIC X(10). */
    @Column(name = "tran_merchant_zip", length = 10)
    private String merchantZip;

    /** TRAN-CARD-NUM PIC X(16). */
    @Column(name = "tran_card_num", length = 16)
    private String cardNumber;

    /** TRAN-ORIG-TS PIC X(26). */
    @Column(name = "tran_orig_ts", length = 26)
    private String originTimestamp;

    /** TRAN-PROC-TS PIC X(26). */
    @Column(name = "tran_proc_ts", length = 26)
    private String processTimestamp;
}
